File: tree/leetcode_Binary_Tree_Level_Order_Traversal.java
/*Binary_Tree_Level_Order_Traversal

Given a binary tree, return the level order traversal of its nodes' values.
(ie, from left to right, level by level).
For example:
Given binary tree {3,9,20,#,#,15,7},
return its level order traversal as:
[
    [3],
    [9,20],
    [15,7]
]
*/

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;


public class leetcode_Binary_Tree_Level_Order_Traversal {

    public static void main(String[] args) {

    }

    public static ArrayList<ArrayList<Integer>> levelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> unvisited = new LinkedList<TreeNode>();
        ArrayList<Integer> tmp = new ArrayList<Integer>();
        unvisited.add(root);

		TreeNode node;
        int count = 1;

        while (!unvisited.isEmpty()) {
            node = unvisited.remove();
            tmp.add(node.val);
            if (node.left != null) {
				unvisited.add(node.left);
			}
            if (node.right != null) {
				unvisited.add(node.right);
			}
            if (--count == 0) {
                result.add(tmp);
                tmp = new ArrayList<Integer>();
                count = unvisited.size();
            }
        }
        return result;
    }
}


class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;

    TreeNode(int val) {
		this.left = null;
		this.right = null;
        this.val = x;
    }
}


/* Python Version
def levelOrderBottom(self, root):
    if not root:
        return []
    
    pipe = [root]
    container = [[root.val]]
    nex = []
    
    while pipe:
        cur = pipe.pop(0)
        if cur.left:
            nex.append(cur.left)
        if cur.right:
            nex.append(cur.right)
        
        if not pipe and nex:
            container.append([ node.val for node in nex ])
            pipe = nex
            nex = []
    return container[::-1]
*/

File: tree/leetcode_Binary_Tree_Level_Order_Traversal.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * binary tree level order traversal
 */

#include <stdio.h>
#include <queue>
#include <vector>

using std::queue;
using std::vector;


/**
 * Definition for a binary tree node.
 */
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
 public:
  vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> levels;
    if (root == NULL) {
      return levels;
    }

    levels.push_back(vector<int> {root->val});

    queue<TreeNode*> q;
    q.push(root);
    vector<TreeNode*> n;

    while (!q.empty()) {
      TreeNode* node = q.front();
      q.pop();
      if (node->left != NULL) {
        n.push_back(node->left);
      }
      if (node->right != NULL) {
        n.push_back(node->right);
      }

      if (q.empty()) {
        if (!n.empty()) {
          vector<int> level;
          for (TreeNode* it : n) {
            level.push_back(it->val);
            q.push(it);
          }
          levels.push_back(level);
        }
        n = vector<TreeNode*> ();
      }
    }
    return levels;
  }
};


int main() {
  Solution sol;

  TreeNode* head = new TreeNode(1);
  head->left = new TreeNode(2);
  head->right = new TreeNode(3);
  head->left->left = new TreeNode(4);
  head->right->right = new TreeNode(5);

  vector<vector<int>> levels = sol.levelOrder(head);

  for (const vector<int> level : levels) {
    for (int val : level) {
      printf("%d ", val);
    }
    printf("\n");
  }
}
